packages/cli/types.d.ts

```diff
@@ -0,0 +1,127 @@
+import webpack from 'webpack';
+
+type Rule = {
+	test: RegExp;
+	[key: string]: any;
+};
+
+type RuleWrapper = {
+	index: number;
+	rule: Rule;
+};
+
+type Loader = {
+	loader: string;
+	options: Record<string, any>;
+};
+
+type LoaderWrapper = {
+	rule: {
+		test: RegExp;
+		[key: string]: any;
+	};
+	ruleIndex: number;
+	loader: Loader;
+	loaderIndex: number;
+};
+
+type Plugin = Record<string, any>;
+
+type PluginWrapper = {
+	index: number;
+	plugin: Plugin;
+};
+
+/**
+ * Basic structure of config
+ */
+export type Config = {
+	resolve: {
+		modules: string[];
+		extensions: string[];
+		alias: string[];
+	};
+	module: {
+		rules: Rule[];
+	};
+	plugins: Plugin[];
+	mode: 'production' | 'development';
+	devtool: string | boolean;
+	output: {
+		filename: string;
+		chunkFilename: string;
+		publicPath: string;
+		[key: string]: any;
+	};
+	devServer?: Record<string, any>;
+	[key: string]: any;
+};
+
+/**
+ * Non-exhaustive list of environment conditions you may want
+ * to check for in your config
+ */
+export type Env = {
+	src: string;
+	dest: string;
+	esm: boolean;
+	sw: boolean;
+	dev: boolean;
+	production: boolean;
+	isProd: boolean;
+	ssr: boolean;
+	prerender: boolean;
+	[key: string]: any;
+};
```

Comment:
Tried to cover the commonly used ones that I've seen. Did I miss any that should be added?